toolchain/check/handle_class.cpp

```diff
@@ -166,6 +167,98 @@ auto HandleClassDefinitionStart(Context& context, Parse::NodeId parse_node)
   return true;
 }
 
+auto HandleBaseIntroducer(Context& /*context*/, Parse::NodeId /*parse_node*/)
+    -> bool {
+  return true;
+}
+
+auto HandleBaseColon(Context& /*context*/, Parse::NodeId /*parse_node*/)
+    -> bool {
+  return true;
+}
+
+auto HandleBaseDecl(Context& context, Parse::NodeId parse_node) -> bool {
+  auto base_type_expr_id = context.node_stack().PopExpr();
+
+  auto class_id =
+      context.node_stack().PeekIf<Parse::NodeKind::ClassDefinitionStart>();
+  if (!class_id) {
+    CARBON_DIAGNOSTIC(BaseOutsideClass, Error,
+                      "`base` declaration can only be used in a class.");
+    context.emitter().Emit(parse_node, BaseOutsideClass);
+    return true;
+  }
+
+  auto& class_info = context.classes().Get(*class_id);
+  if (class_info.base_id.is_valid()) {
+    CARBON_DIAGNOSTIC(
+        BaseRepeated, Error,
+        "Repeated `base` declaration. Multiple inheritance is not permitted.");
+    CARBON_DIAGNOSTIC(BasePrevious, Note,
+                      "Previous `base` declaration is here.");
+    context.emitter()
+        .Build(parse_node, BaseRepeated)
+        .Note(context.insts().Get(class_info.base_id).parse_node(),
+              BasePrevious)
+        .Emit();
+    return true;
+  }
+
+  auto base_type_id = ExprAsType(context, parse_node, base_type_expr_id);
+  if (!context.TryToCompleteType(base_type_id, [&] {
+        CARBON_DIAGNOSTIC(IncompleteTypeInBaseDecl, Error,
+                          "Deriving from incomplete type `{0}`.", std::string);
+        return context.emitter().Build(
+            parse_node, IncompleteTypeInBaseDecl,
+            context.sem_ir().StringifyType(base_type_id, true));
+      })) {
+    base_type_id = SemIR::TypeId::Error;
+  }
+
+  if (base_type_id != SemIR::TypeId::Error) {
+    // For now, we treat all types that aren't introduced by a `class`
+    // declaration as being final classes.
+    auto base_class =
+        context.insts()
+            .Get(context.sem_ir().GetTypeAllowBuiltinTypes(base_type_id))
+            .TryAs<SemIR::ClassType>();
+    if (!base_class ||
+        context.classes().Get(base_class->class_id).inheritance_kind ==
+            SemIR::Class::Final) {
+      CARBON_DIAGNOSTIC(BaseIsFinal, Error,
+                        "Deriving from final type `{0}`. Only abstract and "
+                        "base classes can be derived from.",
```

Comment:
Minor phrasing suggestion, reordering the "can" first.

```suggestion
                        "Deriving from final class `{0}`. Can only derive from `abstract` and `base` classes.",
```

Reply:
Looking at this again, I think "must" works better than "can only" here.